### ISSUE TITLE
Apps that don't support latest

### ISSUE THREAD

**Issue by HugLifeTiZ** [whalehub/ginstall.sh#3]
What would be needed for the apps that don't support latest to support it? I feel like apps that don't support latest defeat the purpose of this application. If I have to go check on what the latest version is, I might as well download the static binary myself anyways.

**Comment by whalehub** [comment]
@HugLifeTiZ I found an elegant solution to this problem which makes the `latest` version argument work with all supported apps (with the same exceptions as the `--check` flag).

I'm planning to publish a new version that includes this change later today.

I'll try to add support for `ffmpeg` and `go` separately later this week.

**Comment by whalehub** [comment]
@HugLifeTiZ I found an elegant solution to this problem which makes the `--check` flag and the `latest` version argument work with all supported apps (including `ffmpeg` and `go`).

I'm planning to publish a new version that includes these changes later today.

**Comment by HugLifeTiZ** [comment]
You know, I was thinking some time back, "why not just parse `--check` in `latest` to deduplicate code?" Again, very nice work, thank you! :)

**Comment by whalehub** [comment]
@HugLifeTiZ You're welcome. The output from `--check` currently looks like this, but you have the right idea:
```sh
whalehub@pdh:~# gget --show-ref github.com/zyedidia/micro
tag     v2.0.4
commit  c5b0c2d41f67048ba859cbf7d7aba6a3b547e1ef
```

Just needs a little bit of `sed` and it can be reused for all applications that are hosted on GitHub:
```sh
whalehub@pdh:~# gget --show-ref github.com/zyedidia/micro | sed '2d;s|tag[[:blank:]]||g;s|v||g'
2.0.4
```
The tricky ones to figure out were `ffmpeg` and `go` since they are hosted elsewhere (i.e. not supported by `gget`), but `grep` and `sed` were also very helpful there.

```sh
whalehub@pdh:~# curl -sSL https://ffmpeg.org/releases/ | grep "tar.gz\"" | sed 's|^.\{62\}||g;s|.tar.gz.*||g' | sort | sed '$!d'
4.2.3
```

```sh
whalehub@pdh:~# curl -sSL https://golang.org/dl/ | grep "span class.*src.tar.gz" | sed 's|^.\{27\}||g;s|.src.*||g'
1.14.4
```

My final solution for `--check` is

```bash
case "$1" in
  "--check" | "-c")
    if [ -z "${GGET_DIR}" ]; then
      echo -e "${MISSING_DEPENDENCY}\n${USAGE_INFORMATION}"
      exit 1
    fi
    case "$2" in
      "ffmpeg")
        RESULT_CHECK="$(curl -sSL https://ffmpeg.org/releases/ | grep "tar.gz\"" | sed 's|^.\{62\}||g;s|.tar.gz.*||g' | sort | sed '$!d')"
        echo -e "The latest version of $2 is v${RESULT_CHECK}."
        exit 0
      ;;

      "go")
        RESULT_CHECK="$(curl -sSL https://golang.org/dl/ | grep "span class.*src.tar.gz" | sed 's|^.\{27\}||g;s|.src.*||g')"
        echo -e "The latest version of $2 is v${RESULT_CHECK}."
        exit 0
      ;;

      *)
        REPO="$(echo "REPO_${2^^}" | sed 's/[-.]/_/g' | sed 's/+/_PLUS/g')"
        RESULT_CHECK="$(gget --show-ref "${!REPO}" 2>/dev/null | sed '2d;s|tag[[:blank:]]||g;s|v||g')"
        if [ -z "${RESULT_CHECK}" ]; then
          echo -e "${UNSUPPORTED_APP}\n${USAGE_INFORMATION}"
          exit 1
        else
          echo -e "The latest version of $2 is v${RESULT_CHECK}."
          exit 0
        fi
      ;;
    esac
  ;;
```

And for `latest`, I came up with

```bash
if [ "$3" == "latest" ]; then
  if [ "$(id -u)" == "0" ] || [ "${DIR_ARG}" == "true" ]; then
    case "$2" in
      "ffmpeg" | "go" )
      ;;

      *)
        APP_VERSION="$(gget --show-ref "${!REPO}" | sed '2d;s|tag[[:blank:]]||g;s|v||g')"
        INSTALL_SUCCESS="The latest version of $2 (v$APP_VERSION) was installed successfully."
      ;;
    esac
  else
    echo -e "${PERMISSION_DENIED}\n${USAGE_INFORMATION}"
    exit 1
  fi
fi
```

and 

```bash
case "$2" in

[...]

    "ffmpeg")
      if [ "$3" == "latest" ]; then
        APP_VERSION="$(curl -sSL https://ffmpeg.org/releases/ | grep "tar.gz\"" | sed 's|^.\{62\}||g;s|.tar.gz.*||g' | sort | sed '$!d')"
        INSTALL_SUCCESS="The latest version of $2 (v$APP_VERSION) was installed successfully."
        FFMPEG_URL="https://johnvansickle.com/ffmpeg/releases/ffmpeg-release-amd64-static.tar.xz"
      else
        FFMPEG_URL="https://www.johnvansickle.com/ffmpeg/old-releases/ffmpeg-${APP_VERSION}-amd64-static.tar.xz"
      fi
      curl -Lf -o "${TMP_DIR_TAR_XZ}" "${FFMPEG_URL}" && \
      tar -xf "${TMP_DIR_TAR_XZ}" -C "${INSTALL_DIR_APP:?}" ffmpeg-"${APP_VERSION}"-amd64-static/"${APP_NAME}" ${TAR_ARGS} --strip-components=1 && \
      chmod +x "${INSTALL_DIR_APP:?}"/"${APP_NAME}" && \
      rm "${TMP_DIR_TAR_XZ}" && \
      echo -e "${INSTALL_SUCCESS}"
    ;;

[...]

    "go")
      if [ "$3" == "latest" ]; then
        APP_VERSION="$(curl -sSL https://golang.org/dl/ | grep "span class.*src.tar.gz" | sed 's|^.\{27\}||g;s|.src.*||g')"
        INSTALL_SUCCESS="The latest version of $2 (v$APP_VERSION) was installed successfully."
      fi
      curl -Lf -o "${TMP_DIR_TAR_GZ}" https://dl.google.com/go/go"${APP_VERSION}".linux-amd64.tar.gz && \
      rm -rf "${INSTALL_DIR_GO:?}"/"${APP_NAME}" && \
      tar -xf "${TMP_DIR_TAR_GZ}" -C "${INSTALL_DIR_GO:?}" ${TAR_ARGS} && \
      rm "${TMP_DIR_TAR_GZ}" && \
      echo -e "${INSTALL_SUCCESS}"
    ;;
esac

[...]
```

I don't know if I'll ever learn a programming language but solving problems in Bash is a lot of fun. 😄

**Comment by whalehub** [comment]
@HugLifeTiZ Just published [v2.0.0](https://github.com/whalehub/ginstall.sh/releases/tag/v2.0.0) which includes the fix for this. My release numbers use semantic versioning, which means that this is a breaking release. Please make sure to read the full changelog and revamped `--help` flag output. 👍
